### PR TITLE
ECS infrastructure with the Flagsmith task definition

### PR DESCRIPTION
**Pull request by deltacodepl**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?

## Changes

Added "flagsmith-on-ecs" subproject with Terraform code that spins up ECS Fargate test cluster with the Flagsmith basic task definition.  
The project provides instant access to the Flagsmith instance under specified domain by adding the flagsmith prefix such as ```flagsmith.yourdomain.com```

### How to test this project

1. Register AWS Route53 Hosted Zone \
  for example ```yourdomain.com```

2. Generate certificate with AWS Certificate Manager \
  with ```*.yourdomain.com``` pattern

3. Define your variables such as certificate arn, hosted zone domain, and other desired settings in **terraform.tfvars** file 
    ```bash
    route53_hosted_zone = "yourdomain.com"
    certificate_arn = "arn:aws:acm:"
    ```
4. Generate plan for infrastructure ```terraform plan -out flagsmith.tfplan```
5. Apply infrastructure by running ```terraform apply flagsmith.tfplan```

### CODE REVIEW

**Comment by khvn26** [comment]
This is awesome, @deltacodepl, thank you so much for your contribution! Bonus points for the docs 🔥 

There are formatting issues detected in CI, any chance you could look into it?

**Comment by deltacodepl** [comment]
Sure thing,
Thank you for your appreciation, Kim

File: examples/flagsmith-on-ecs/terraform/04_loadbalancer.tf [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    interval            = 5
+    matcher             = "200"
+  }
+}
+
+
+# Listener (redirects traffic from the load balancer to the target group)
+resource "aws_alb_listener" "ecs-alb-http-listener" {
+  load_balancer_arn = aws_lb.production.id
+  port              = "443"
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/04_loadbalancer.tf]: I think that the load balancer SG allows 443 and 80, but there's only a 443 listener.

Suggest either:
- 443 only on the listener, 443 only on the load balancer SG
- 443 on the listener that forwards to the workload, 80 on the listener that redirects to 443, 443 and 80 on the load balancer SG (probably with a comment to say that 80 is only used for redirecting to 443)

File: examples/flagsmith-on-ecs/terraform/04_loadbalancer.tf [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  protocol          = "HTTPS"
+  ssl_policy        = "ELBSecurityPolicy-2016-08"
+  certificate_arn   = var.certificate_arn
+  depends_on        = [aws_alb_target_group.default-target-group]
+
+  default_action {
+    type             = "forward"
+    target_group_arn = aws_alb_target_group.default-target-group.arn
+  }
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/04_loadbalancer.tf]: We've seen issues on ALB+ECS workloads, where requests like `curl -k https://[load balancer ip]` get to the load balancer and are forwarded, and generate lots of `DisallowedHost` errors from Django.

For this reason, think it is preferable here to have an action that matches on the expected host name, and forward that to the ALB target group. Then the default action is 404. As it happens, here is me suggesting this as an improvement in the CDK: https://github.com/aws/aws-cdk/issues/25434.

File: examples/flagsmith-on-ecs/terraform/04_loadbalancer.tf [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+
+# Listener (redirects traffic from the load balancer to the target group)
+resource "aws_alb_listener" "ecs-alb-http-listener" {
+  load_balancer_arn = aws_lb.production.id
+  port              = "443"
+  protocol          = "HTTPS"
+  ssl_policy        = "ELBSecurityPolicy-2016-08"
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/04_loadbalancer.tf]: This should probably be `ELBSecurityPolicy-TLS13-1-0-2021-06`, with a comment to say that the user should ensure it is appropriate for their usage, and use a newer one if available when they are deploying, see https://docs.aws.amazon.com/elasticloadbalancing/latest/application/create-https-listener.html#describe-ssl-policies.

File: examples/flagsmith-on-ecs/terraform/05_iam.tf [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      # used for ECS Exec
+      {
+        "Effect" : "Allow",
+        "Action" : [
+          "ssmmessages:CreateControlChannel",
+          "ssmmessages:CreateDataChannel",
+          "ssmmessages:OpenControlChannel",
+          "ssmmessages:OpenDataChannel"
+        ],
+        "Resource" : "*"
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/05_iam.tf]: Does this have to be `*`?

File: examples/flagsmith-on-ecs/terraform/05_iam.tf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+resource "aws_iam_role_policy" "ecs_task" {
+  name = "${var.app_name}-ecs-task-policy"
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Effect   = "Allow"
+        Action   = ["secretsmanager:GetSecretValue", "ssm:GetParameters"],
+        Resource = ["arn:aws:ssm:${var.region}:${local.AWS_ACCOUNT_ID}:parameter/${var.app_name}/"]
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/05_iam.tf]: This has a `secretsmanager:...` permission action, but the resource is `arn:aws:ssm:...`, which looks wrong to me.

File: examples/flagsmith-on-ecs/terraform/07_route53.tf [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  records = [aws_lb.production.dns_name]
+}
+
+resource "aws_route53_record" "www" {
+  zone_id = data.aws_route53_zone.selected.zone_id
+  name    = "www.${aws_route53_record.subdomain.name}"
+  type    = "CNAME"
+  ttl     = "300"
+  records = [aws_route53_record.subdomain.name]
+}
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/07_route53.tf]: I think this is slightly surprising behaviour from an example, that it includes the `www` subdomain. Suggest this is noted in the readme, and maybe this should be replaced with more guidance about how to add extra domains, and what that means for:
- DNS
- the certificate
- the ALLOWED_HOSTS setting in Django

File: examples/flagsmith-on-ecs/terraform/07_route53.tf [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  private_zone = false
+}
+
+resource "aws_route53_record" "subdomain" {
+  zone_id = data.aws_route53_zone.selected.zone_id
+  name    = "${var.app_name}.${data.aws_route53_zone.selected.name}"
+  type    = "CNAME"
+  ttl     = "300"
+  records = [aws_lb.production.dns_name]
+}
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/07_route53.tf]: Is there a reason not to use an Alias record here?

File: examples/flagsmith-on-ecs/terraform/11_rds.tf [hunk 8]
```diff
@@ -0,0 +1,27 @@
+
+resource "aws_db_subnet_group" "production" {
+  name = "main"
+  # subnet_ids = module.vpc.public_subnets
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/11_rds.tf]: I think this line should probably go away, as I think we'd always want to encourage using private subnets for a database.

File: examples/flagsmith-on-ecs/terraform/11_rds.tf [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+
+resource "aws_db_instance" "production" {
+  identifier              = "production"
+  db_name                 = var.rds_db_name
+  username                = var.rds_username
+  password                = aws_ssm_parameter.db_password.value
+  port                    = "5432"
+  engine                  = "postgres"
+  engine_version          = "11.12"
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/11_rds.tf]: Is there a reason not to use a newer Postgres?

File: examples/flagsmith-on-ecs/terraform/auto_scaling.tf [hunk 10]
```diff
@@ -0,0 +1 @@
+## Fargate autoscaling
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/auto_scaling.tf]: Is this to be removed, or should there be some resources here?

File: examples/flagsmith-on-ecs/terraform/policies/ecs-host-role-policy.json [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        "elasticloadbalancing:DeregisterTargets",
+        "cloudwatch:*",
+        "s3:*",
+        "rds:*",
+        "logs:*",
+        "elasticache:*",
+        "secretsmanager:*",
+        "ssm:*"
+      ],
+      "Resource": "*"
```
Comment (plumdog) [inline examples/flagsmith-on-ecs/terraform/policies/ecs-host-role-policy.json]: This is very permissive. Could this be pruned?

**Review by plumdog — COMMENTED** [review]
I think we need to be careful about what is being promised with this code.

If it is intended to be something that a user - a beginner with Terraform and AWS - could pick up and use to deploy their production environment, then I think the bar is very high. We'd need to accommodate lots of different options (things like the database sizing, database snapshotting, fargate resource allocation options and autoscaling, VPC and CIDR options etc) and absolutely need to be very careful with the IAM policies to reduce them to be least privilege.

Or, we could make very clear that this is a prove-to-yourself-and-your-boss-that-it-works example, but nothing more. Remove all references to "production" from this code and put a big warning at the top of the README to explain that this is not production ready, and that a user who is familiar with Terraform and AWS should use this as a reference to build their own solution.

**Comment by deltacodepl** [comment]
Hi @plumdog 
Thank you for the review
I have cleaned up the code, tightened IAM permissions, and removed unnecessary elements based on all your suggestions.

**Review by plumdog — APPROVED** [review]
LGTM :+1: 

Thanks for applying the changes, this looks great, thanks for the hard work.

I must confess I haven't run it though (I feel like someone Flagsmith-side should actually run this code at least once before we merge... [looks around for volunteers...])

File: flagsmith-on-ecs/terraform/variables.tf [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+# logs
+
+variable "log_retention_in_days" {
+  type    = number
+  default = 7
+}
+
+# specify dev/prod django settings
+variable "settings_module" {
+  type = map(string)
```
Comment (khvn26) [inline flagsmith-on-ecs/terraform/variables.tf]: Not sure what's the logic here as the ECS definition only uses the module specified under the `"dev"` key. Could we
1. Get rid of the map and use plain `string` for this value and
2. Have a default of `app.settings.production` which will be used by 99% of the users anyway?

File: flagsmith-on-ecs/terraform/variables.tf [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+variable "health_check_path" {
+  type        = string
+  description = "Health check path for the default target group"
+  default     = "/health/"
+}
+
+# ecs fargate
+
+
+variable "docker_image_url" {
```
Comment (khvn26) [inline flagsmith-on-ecs/terraform/variables.tf:34]: This definitely can have a default of `flagsmith/flagsmith:latest`.

File: flagsmith-on-ecs/terraform/variables.tf [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+  default     = ""
+}
+
+variable "cpu" {
+  type = number
+}
+
+variable "memory" {
+  type = number
+}
```
Comment (khvn26) [inline flagsmith-on-ecs/terraform/variables.tf]: I feel these are too permissive given that Fargate is used. Could be mitigated by having sensible defaults which correspond to Fargate's templates: cpu = 256 and memory = 512.

**Review by khvn26 — DISMISSED** [review]

**Comment by khvn26** [comment]
It works, with some minor comments I've added. @deltacodepl I'll leave those for a bit of time in case you want to address them. Otherwise, I'll merge the PR and fix those myself. Again, many thanks for your work! Great job! 👍

**Comment by dabeeeenster** [comment]
Yep thanks again for this @deltacodepl

**Review by khvn26 — APPROVED** [review]

**Pull request merged**